### ISSUE TITLE
C# memory leak with multiple channels

### ISSUE THREAD

**Issue by jskeet** [grpc/grpc#14021]
- Grpc.Core v1.8.0 and 1.8.3 behave very badly; Grpc.Core v1.7.3 is somewhat better (details below)
- Windows 10

Simple repro using `Google.Cloud.Speech.V1` library just for the generated protos and gRPC code:

```csharp
using Google.Apis.Auth.OAuth2;
using Google.Cloud.Speech.V1;
using Grpc.Auth;
using Grpc.Core;
using System;

public class Test
{
    public static void Main(string[] args)
    {
        GoogleCredential googleCredential = GoogleCredential.GetApplicationDefault();
        ChannelCredentials channelCreds = googleCredential.ToChannelCredentials();
        for (int i = 0; i < 100; ++i)
        {
            var channel = new Channel(SpeechClient.DefaultEndpoint.Host, channelCreds);
            var speech = new Speech.SpeechClient(channel);
            var call = speech.StreamingRecognize();
            call.RequestStream.CompleteAsync().Wait();
            call.ResponseStream.MoveNext(System.Threading.CancellationToken.None).Wait();
            channel.ShutdownAsync().Wait();
            Console.WriteLine($"{i}: {GC.GetTotalMemory(true)}");
        }
    }
}
```

Project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Google.Cloud.Speech.V1" Version="1.0.0" />
    <PackageReference Include="Grpc.Core" Version="1.8.3" />
  </ItemGroup>
</Project>
```

Output when using Grpc.Core v1.7.x:

```text
0: 376984
1: 382392
2: 382392
3: 382960
4: 382960
5: 382960
6: 382960
7: 382960
8: 382960
9: 382960
(stable)
```

When using Grpc.Core v1.8.x:

```text
0: 461312
1: 553512
2: 640344
3: 728456
4: 815368
5: 902816
6: 990232
7: 1077648
8: 1164728
9: 1250880
(keeps increasing)
```

Now even 1.7.x isn't perfect: if you comment out the call to `call.ResponseStream.MoveNext(System.Threading.CancellationToken.None).Wait();`, there's still a small leak, although it's much better than with 1.8.x:

```text
0: 376976
1: 382504
2: 383136
3: 383280
4: 383424
5: 383648
6: 383792
7: 383936
8: 384080
9: 384224
```

Using a single channel instead of multiple channels avoids the leak even with 1.8.x, and even without calling `MoveNext()` to observe the end of the stream.

**Comment by jtattermusch** [comment]
@mehrdada  we should fix this for v1.9.x release.

**Comment by jtattermusch** [comment]
Possibly related https://github.com/grpc/grpc/issues/13941

**Comment by jtattermusch** [comment]
I believe a the problem you are describing with 1.7.0 (when MoveNext is omitted) is https://github.com/grpc/grpc/issues/8451.

I am looking into the 1.8.x leak.

**Comment by jtattermusch** [comment]
I know what's going on here:
- After in v1.8.x I've introduced to PRs for pooling BatchContextSafeHandle and RequestCallContextSafeHandle objects: (https://github.com/grpc/grpc/pull/13476, https://github.com/grpc/grpc/pull/13521).
- the cleanup of the thread local pools from DefaultObjectPool seems not to be working well after the threads finish (the expectation was that after the thread finishes, the thread local objects stop being reachable at they will get garbage collected, but that might not be happening).
- GrpcEnviroment is a reference-counted singleton (channel and server creation increment the refcount, channel and server shutdown decrement). So in this particular example, because there's only 1 channel at a time and it gets created and shutdown multiple times, in each loop the GrpcEnviroment is shutdown and initialized again (which involves creation of GrpcThreadPool). 
- the combination of the previous factors means that each loop of the test creates several new threads and leaks the thread local objects for them and they are never cleaned up, so memory use keeps growing.

I'm still thinking what would be the best way to fix the problem (rolling back the 2 PRs that are not independent seems a bit tricky).

**Comment by mehrdada** [comment]
@jtattermusch sounds good. Do we need to delay the branch cut as well as the release? I was hoping to cut the branch, and we can ship a separate RC for C# a bit later, as this gives more time for RC of other languages?

**Comment by mehrdada** [comment]
Ok. Cutting the branch now...

**Comment by jtattermusch** [comment]
Fixed by #14113.
The fix will be released in v1.9.x-pre2

**Comment by jtattermusch** [comment]
@jskeet   the v1.9.0-pre2 packages have been pushed.
